### PR TITLE
Fix ArrayInput Adding Previously Removed Item

### LINKED ISSUE #7385: Adding an item to ArrayInput on Edit after removing one brings back the same item that was removed
**What you were expecting:**
Removing an item from ArrayInput should really remove it.

**What happened instead:**
After removing an item and adding a new one, the previosuly removed item is back.

**Steps to reproduce:**
Start here: [codesandbox.io/s/happy-germain-877ypu](codesandbox.io/s/happy-germain-877ypu)
1. Create a new post using the floating + button in bottom right,, fill in the required fields and make sure it has a backlink added. Click on *Save and edit*.
2. When Edit view has opened, click on "Miscellaneous" tab to edit backlinks.
3. Remove the backlink.
4. Add a backlink.
5. Observe, the backlink you previosly removed is now back.
I have added a video for clarification:

https://user-images.githubusercontent.com/20572374/158165800-91da4c4c-47c9-438b-a847-ad27ea0c1010.mp4

**Other information:**
There is a workaround which fixed the problem for me:

```tsx
const AddItemButton = ({ onClick, ...rest }: ButtonProps) => {
  const { change, getFieldState } = useForm();
  return (
    <Button
      label="ra.action.add"
      {...rest}
      onClick={(e) => {
        const fieldState = getFieldState("formArrayFieldSource");
        const nextIndex = fieldState?.value && Array.isArray(fieldState.value) ? fieldState.value.length : 0;
        onClick && onClick(e);
        change(`formArrayFieldSource[${nextIndex}]`, {
          ...getEmptyArrayItem()
        });
      }}>
      <AddCircleOutlineRoundedIcon />
    </Button>
  );
};
```

where `getEmptyArrayItem` is a function which returns an empty array item, such that it doesn't have any undefined members.
The idea behind it is to manually make sure the newly added array item is really **new**.

**Environment**

* React-admin version: 3.19.10 (tried with *3.19.3* too but locally, same problem)
* Last version that did not exhibit the issue (if applicable): N/A
* React version: 16 locally, but the sandbox uses 17
* Browser: Chrome latest

### PR DESCRIPTION
**Pull request by djhi**
Fixes #7385

### CODE REVIEW

File: packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.spec.tsx [hunk 1]
```diff
@@ -348,6 +348,58 @@ describe('<SimpleFormIterator />', () => {
...
+                        source="emails"
+                        defaultValue={[{ email: 'test@marmelab.com' }]}
+                    >
+                        <SimpleFormIterator>
+                            <TextInput
+                                source="email"
+                                label="CustomLabel"
+                                defaultValue="default@marmelab.com"
+                            />
+                        </SimpleFormIterator>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.spec.tsx:365]: could you also include an input with no default value?

File: packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.tsx [hunk 2]
```diff
@@ -56,21 +55,21 @@ export const SimpleFormIterator = (props: SimpleFormIteratorProps) => {
...
+                    }
+                });
+            }
+            append(defaultValue);
         },
-        [append]
+        [append, children]
     );
 
-    // add field and call the onClick event of the button passed as addButton prop
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.tsx:64]: Why did you remove this part?

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.tsx (hunk 2 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.tsx:64]: Because it was unnecessary and actually induced an issue by adding an item twice

File: packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.spec.tsx (hunk 1 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.spec.tsx:368]: Done

**Review by fzaninotto — APPROVED** [review]

**Review by fzaninotto — CHANGES_REQUESTED** [review]

**Review by septentrion-730n — APPROVED** [review]

File: packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.tsx (hunk 2 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.tsx:64]: restored

**Review by fzaninotto — APPROVED** [review]
